trusted_application_demo/client_application/aes_certify_messaging/subscriber/aes_subscriber.cpp

```diff
@@ -0,0 +1,186 @@
+    // Used to store the published message
+    ca_teec_exit_on_failure(&ctx, &sess, ca_teec_allocate_shared_memory(&ctx, &clear_message_shm,
+      SUBSCRIBER_MAX_MESSAGE_SIZE, TEEC_MEM_INPUT));
+    // Used to store the digest of the message signed by AES
+    ca_teec_exit_on_failure(&ctx, &sess, ca_teec_allocate_shared_memory(&ctx, &sha_shm,
+      HMAC_API_SHA1_SIZE_BYTE, TEEC_MEM_INPUT));
+
+    // Used to authenticate the secret key send. Store the RSA digest of AES key
+    ca_teec_exit_on_failure(&ctx, &sess, ca_teec_allocate_shared_memory(&ctx, &sha1_rsa_signed_shm,
+      RSA_API_SIZE_ENCRYPTED_BYTE, TEEC_MEM_INPUT));
+    // Used to store the incomming RSA encrypted AES key
+    ca_teec_exit_on_failure(&ctx, &sess, ca_teec_allocate_shared_memory(&ctx,
+      &aes_encrypted_key_shm, RSA_API_SIZE_ENCRYPTED_BYTE, TEEC_MEM_INPUT));
+  }
+
+  void waiting_for_secret(std::string service)
+  {
+    auto client_ = this->create_client<GetSymmSecret>(service);
+
+    while (!client_->wait_for_service(std::chrono::seconds(1))) {
+      if (!rclcpp::ok()) {
+        RCLCPP_ERROR(this->get_logger(), "Interrupted before service came !");
+        ca_teec_close_session(&ctx, &sess);
+        exit(0);
+      }
+      RCLCPP_INFO(this->get_logger(), "Waiting for the secret key.");
+    }
+
+    auto request = std::make_shared<GetSymmSecret::Request>();
+    bool validated = false;
+    while (!validated) {
+      auto answer_service = client_->async_send_request(request);
+      if (rclcpp::spin_until_future_complete(this->get_node_base_interface(),
+        answer_service) != rclcpp::executor::FutureReturnCode::SUCCESS)
+      {
+        RCLCPP_ERROR(this->get_logger(), "Service could not deliver the secret... Exiting");
+        ca_teec_close_session(&ctx, &sess);
+        exit(-1);
+      }
+
+      // Fill the shared memory with the AES secret encrypted by RSA
+      auto result = answer_service.get();
+      uint8_t * aes_secret = reinterpret_cast<uint8_t *>(aes_encrypted_key_shm.buffer);
+      for (uint32_t i = 0; i < result->secret.size() && i < aes_encrypted_key_shm.size; i++) {
+        aes_secret[i] = result->secret[i];
+      }
+
+      // Fill the shared memory with the RSA signature of the digested key
+      uint8_t * rsa_digest = reinterpret_cast<uint8_t *>(sha1_rsa_signed_shm.buffer);
+      for (uint32_t i = 0; i < result->sha.size() && i < sha1_rsa_signed_shm.size; i++) {
+        rsa_digest[i] = result->sha[i];
+      }
+
+      // Check that the digest is valid
+      validated = ca_rsa_verify(&sess, &aes_encrypted_key_shm, aes_encrypted_key_shm.size,
+          &sha1_rsa_signed_shm, sha1_rsa_signed_shm.size);
+    }
+  }
+
+  void topic_callback(const AesMessageCertify::SharedPtr msg)
+  {
+    // Copy the content of the message into the shared memory location
+    uint8_t * clear_cast = reinterpret_cast<uint8_t *>(clear_message_shm.buffer);
+    for (uint32_t i = 0; i < msg->message.size() && i < clear_message_shm.size; i++) {
```

Comment:
`std::vector.size()` returns `size_t` and not `uint32_t` http://en.cppreference.com/w/cpp/container/vector/size (same for every size() call)